main.vala

```diff
@@ -178,334 +179,225 @@ class Vls.Server {
+                    showMessage (client, @"Failed to parse meson targets: $(e.message)", MessageType.Error);
                 }
             }
-        } catch (Error e) {
-            debug (@"Error adding files: $(e.message)");
-        }
-    }
-
-    void default_analyze_build_dir (Jsonrpc.Client client, string root_dir) {
-        try {
-            add_vala_files (File.new_for_path (root_dir));
-        } catch (Error e) {
-            debug (@"Error adding files $(e.message)n");
-        }
-    }
-
-    void initialize (Jsonrpc.Server self, Jsonrpc.Client client, string method, Variant id, Variant @params) {
-        init_params = parse_variant<InitializeParams> (@params);
-
-        string root_path = init_params.rootPath != null ? init_params.rootPath : init_params.rootUri;
-        debug (@"Root path is $root_path");
-
-        string? meson = findFile (root_path, "meson.build");
-        if (meson != null) {
-            string? ninja = findFile (root_path, "build.ninja");
-
-            if (ninja == null) {
-                // TODO: build again
-                // ninja = findFile (root_path, "build.ninja");
+        } else {
+            try {
+                // we don't support anything else, so just create a default target
+                builds.add (new SimpleTarget (root_path));
+            } catch (Error e) {
+                showMessage (client, @"Failed to add simple target for project: $(e.message)", MessageType.Error);
+                debug (@"Failed to add SimpleTarget: $(e.message)");
             }
+        }
 
-            // test again
-            if (ninja != null) {
-                debug ("Found meson project: %s\nninja: %s", meson, ninja);
-                meson_analyze_build_dir (client, root_path, Path.get_dirname (ninja));
-            } else {
-                debug ("Found meson.build but not build.ninja: %s", meson);
-                string? configDir = mesonConfigure (client, meson);
-                if (configDir != null) {
-                    meson_analyze_build_dir (client, root_path, configDir);
+        // sanity checking
+        var text_documents = new HashMap<string, TextDocument> ();
+        var build_targets_to_remove = new ArrayList<BuildTarget> ();
+        foreach (var build_target in builds) {
+            foreach (var compilation in build_target) {
+                foreach (var document in compilation) {
+                    if (text_documents.has_key (document.uri)) {
+                        var other_document = text_documents[document.uri];
+                        var target1 = other_document.compilation.parent_target;
+                        var target2 = document.compilation.parent_target;
+                        debug (@"[$method] the same text document $(document.uri) appears twice in $(target1) and $(target2)!");
+                        debug (@"[$method] will remove $(target2)");
+                        build_targets_to_remove.add (target2);
```

Comment:
could this break other functionality which depends on all targets being present in a build?

Reply:
Maybe, but for now it's the best option I could come up with.